### ISSUE TITLE
Fan-in not working

### ISSUE THREAD

**Issue by CharlieC3** [gocd/gocd#2225]
##### Issue Type
- Bug Report
##### Summary

In the context of fan-ins, I'm seeing a pipeline dependent on multiple previous pipelines triggering before all dependent pipelines finish running.
###### Basic environment details

- Go Version: `16.4.0`
- JAVA Version: `1.8`
- OS: `Linux, amd64, 3.10.0-229.1.2.el7.x86_64`
##### Steps to Reproduce
1. Create pipeline 'A'
2. Create many more pipelines which depend on 'A'
3. Create pipeline 'Z', which depends on all pipelines created in Step 2. 'Z' does not DIRECTLY depend on 'A', but only transitively through the other pipelines.
4. Trigger pipeline 'A'
##### Expected Results
1. Pipeline 'A' runs and completes.
2. All pipelines following 'A' start.
3. Pipelines following 'A' complete in a staggered fashion.
4. Pipeline 'Z' runs ONLY after all previous pipelines in step 2 are completed.
##### Actual Results
1. Pipeline 'A' runs and completes.
2. All pipelines following 'A' start.
3. Pipelines following 'A' complete in a staggered fashion.
4. Pipeline 'Z' kicks off early, before all previous pipelines can finish.
5. More pipelines following 'A' finish.
6. Pipeline 'Z' kicks off again, using the recently completed pipelines as materials.

Steps 4 and 5 repeat until all pipelines following 'A' have finished.

**Comment by benzvan** [comment]
I can see the utility of this and it speaks to the current handling of upstream dependencies. If the pipeline execution is set to "on success", any material listed will trigger the pipeline. This sounds like a need for an additional option beyond "manual" and "on success" to accommodate only-when-all-materials-pass since the behavior you describe as actual results is probably the way most users use materials. For example, with a code material and a property material, changes in either one should trigger a build.

**Comment by mynkow** [comment]
I also think that this is a bug but probably this is by design because I have this problem from day of the project. 

There is no need of additional checkbox IMO.

**Comment by Sinstraliz** [comment]
This is something that we have been missing from the start. It will be nice to have because currently we have to create a workarounds for this and chain pipelines that are not related to each other but they are related to another pipeline.

**Comment by smalldek** [comment]
This is a bug. You can see in the documentation https://docs.go.cd/current/advanced_usage/fan_in.html that it need to wait until the end of both previous pipelines. I will try to test with an older version because, for me, it worked a couple of month.

**Comment by mynkow** [comment]
Any idea if this is actually a bug or it was intentionally reworked this way?

**Comment by Stono** [comment]
Hey all,
I've experienced this today - I don't think this is a bug (doesn't mean I like the behaviour).

In a [Fan In](https://docs.gocd.io/current/advanced_usage/fan_in.html) scenario, the fan in pipeline will be triggered by the __latest__ green version of the upstream dependencies.

I think a lot of us are expecting the behaviour where the Fan In pipeline is only trigger when the upstream builds are both green __from the same version__.

To give a literal example, taking the following diagram from the gocd:
![Pipeline](https://docs.gocd.io/current/resources/images/fanin_2_1.png)

__Given__ c1:1, c2:1, c3:1 builds are green
__When__: c1:2 builds and goes green
__And__: c2:2 builds and goes green
__And__: c3:2 builds and goes __RED__

_What we get now_:
__Then__: c4:2 builds using c2:2 and c3:1 (the latest green of each)

_What we should get, or should at least have the option of_:
__Then__: c4:2 does not build, c4:2 should only trigger when c2:2 and c1:2 are both green

**Comment by rajiesh** [comment]
@Stono 

For the scenario you explained in your comment, the reason for the said behavior is because fan-in resolution goes back to the SCM commit which triggered the first pipeline C-1. Let me try to explain with the same scenario,

1. Pipeline C-1 triggered first by the SCM commit A
2. C-1:1 is Green now
3. C-1:1 will now trigger C-2:1 and C-3:1 and both goes Green
4. Now fan-in kicks in and PACKAGE:1 will be triggered by C-2:1 and C-3:1
5. Now all the pipelines are Green
6. At this stage for some reason if user triggered C-1 manually(with the same commit A) and C-1:2 completed successfully(Green)
7. C-1:2 will trigger C-2:2 and C-3:2
8. Considering that C-2:2 went green but C-3:2 went Red, the PACKAGE:2 will still get triggered because it had a successful fan-in resolution in the upstream SCM(A)->C-1:2->C-2:2 and C-3:1

Now consider a change in the scenario. If the steps after Step 5 are as below the behavior will change,

6. User makes a new commit to SCM say B
7. Commit B will now trigger C-1:2 and it goes Green
8. C-1:2 will trigger C-2:2 and C-3:2
9. Say C-2:2 goes Green and C-3:2 goes Red

Now PACKAGE:2 will not get triggered because the fan-in resolution will not be satisfied since the C-3:2 failed for SCM commit B

Let know if this is not the behavior you are observing or if you see any concerns in this behavior

**Comment by MattAlex151** [comment]
Hi all,

We're seeing the same issue and also understand from the documentation that this is not intended behaviour. 

We have four pipelines, A, B, C, and D. Pipeline A builds and kicks of B+C simultaneously, and I want to wait for the latest green, passing B+C before starting D, so I set B+C as a material to D and start A. Then when B finishes (but C is still going) D starts with New B and Old C, then when C finishes, D starts with New B and New C. 

As far as I can tell/read, D should have only started once with both New B + New C. Thanks!

**Comment by Stono** [comment]
@rajiesh could we perhaps have some diagrams explaining this in more detail, I *think* i understand the situation, and even if I do understand it, I'm not sure I agree with it?

**Comment by slavina-rumenova** [comment]
Hi all, 
is there an update on this issue? 
I am facing it as well as a part of a Deployment pipeline, so it's not possible to leave it execute the Fade in pipeline multiple times. Has anybody managed to find a workaround?
Will this be fixed in the foreseeable future?

Cheers

**Comment by ouzo66** [comment]
Hi all,

as a short example:
[gocd_test.txt](https://github.com/gocd/gocd/files/1007129/gocd_test.txt)

By triggering this by an git push, everything runs fine. - OK

By manual Triggering the merge and the following steps are running twice. - NOK

**Comment by SubNader** [comment]
Hello all,

Is there any updates regarding this issue?

Thanks.
